docs/features/role-based-access-control/high-level-design.md

```diff
@@ -0,0 +1,81 @@
+to custom actions that might be subject to RBAC: for instance, if publishing a project is a custom action that should be
+subject to RBAC, then it is up to the controller action, service method, or wherever we need to do this, to check that 
+the user performing the request has the required role (e.g. `Owner`) to perform that action.
+
+Currently, these are the roles required to perform the following actions in the platform:
+
+| Actions / Roles    | Scenario Viewer | Scenario Contributor | Scenario Owner | Project Viewer | Project Contributor | Project Owner | Organization Viewer | Organization Contributor | Organization Owner | Platform admin |
+|--------------------|:---------------:|:--------------------:|:--------------:|:--------------:|:-------------------:|:-------------:|:-------------------:|:------------------------:|:------------------:|:--------------:|
+| Edit project       | n               | n                    | n              | n              | Y                   | Y             | n                   | Y                        | Y                  | Y              |
+| Publish my project | n               | n                    | n              | n              | n                   | Y             | n                   | n                        | Y                  | Y              |
+| Delete project     | n               | n                    | n              | n              | n                   | Y             | n                   | n                        | Y                  | Y              |
+| Delete scenario    | n               | n                    | Y              | n              | n                   | n             | n                   | n                        | Y                  | Y              |
+
+## Association between roles and users for resources
+
+The association between roles and users for resources will be stored in database tables in the `marxan-api` database,
+using one database table for each resource:
+* `users_organizations` will store the roles of users inside organizations.
+* `users_projects` will store the roles of users inside projects.
+* `users_scenarios` will store the roles of users inside scenarios.
+
+These tables should contain columns for `user_id`, `role` and the resource id (`organization_id`, `project_id` and 
+`scenario_id`, respectively).
+
+## Architecture
+
+A main module `access-control` should be created, which will be the main point-of-contact for checking permissions. 
+Additionally, we will have specific modules for checking permissions for each of the resources supported:
+* `organizations-acl`: module that controls access to organizations;
+* `projects-acl`: module that controls access to projects;
+* `scenarios-acl`: module that controls access to scenarios;
+* `solutions-acl`: module that controls access to solutions;
+
+This `access-control` module will expose a public service, and this is the point of entry that should be used by other 
```

Comment:
While I agree with the suggestion to have a single entry point for authorization questions we need to ask, I invite to consider using a different interface whereby instead of asking "Does user U have role R on resource M?" we would ask "Can user U do action A on resource M?".

In practice, I think just checking the role could work for the moment (and it can even be proxied via `canUserDoThis()`-style interface, if we wanted to provide more descriptive semantics). From my experience (greybeard alert!) a system that assigns roles but checks permissions is more flexible, and all in all I tend to prefer some more complexity upfront that can allow to deal with "yes. but about this edge case?" as development progresses (not necessarily in a future phase, even to wrap up the current one). A simpler setup may still well turn out to be just enough for the time being, though.

Reply:
@hotzevzl isn't it something we already did?

```
async canPublish(userId: string, projectId: string): Promise<boolean>
```
(instead of checking role)

Reply:
While I agree this approach is more flexible, one could also argue it's less reusable, harder to maintain, and even redundant, in some cases. For every new action in the platform subject, we will have to add a new method to check permissions for that action. The purpose of having roles (IMO) is exactly to avoid having to check each action individually. Each role has a set of actions they can do (identified in the permissions matrix), and that way we can take advantage of the existing methods to check permissions for a role, instead of having to create a new method for every new action.

However, I'll accept the suggestion, mostly because of the greybeard card :sweat_smile: if flexibility is key, then this approach is more adequate in this case.

Reply:
@ikas there is no silver bullet. As you mentioned, this approach has it downsizes but on the other hand, leaking the  roles lead to multiple modules/places/services to check permissions on their own - and it also will eventually become duplicated many times, which makes the changes much more harder to maintain when compared to proposed solution. Again, there is no silver bullet ;)

Reply:
> @hotzevzl isn't it something we already did?
> > async canPublish(userId: string, projectId: string): Promise<boolean>

yep, that's what I meant.

ACL consumers should not need to know about roles, but about subjects and permissions. To be fair, for the time being these abstractions and separations of concerns may be more academic than pragmatic, but depending on the implementation, checking permissions vs checking roles could be a negligible computational overhead.

Things would be quite different if, say, platform admins had an app UI where they could reconfigure which roles are granted which permissions - in this case, as long as no new permissions are added, the "grant roles, check permissions" strategy would allow to reshape roles without code changes. But indeed, this is not a requirement for the time being.